Add "~use_magnetic_field_msg" param.

mikepurvis opened the pull request:
This allows the user to subscribe to the /imu/mag topic as a
sensor_msgs/MagneticField rather than a geometry_msgs/Vector3Stamped.
The default for now is false, which preserves the legacy behaviour via a
separate subscriber which converts Vector3Stamped to MagneticField and
republishes.

Addresses #22. Requires further testing before merge.

mintar commented:
Thanks for taking the initiative on this! I'll hopefully get around to testing this on Monday; in the meantime, here are a couple of comments on the code:

I think it would be more straightforward not to republish the mag message inside the node, and put that into a separate node instead (could be a Python three-liner). That part isn't finished anyway (you've set up `mag_republisher_`, but never use it, right?).

Another small thing: You've fixed a lot of whitespace errors; it would be easier to see what the important changes are if you had put the whitespace fixes into a separate commit. Sorry for the nitpicking.

mikepurvis committed:
Add "~use_magnetic_field_msg" param.

This allows the user to subscribe to the /imu/mag topic as a
sensor_msgs/MagneticField rather than a geometry_msgs/Vector3Stamped.
The default for now is false, which preserves the legacy behaviour via a
separate subscriber which converts Vector3Stamped to MagneticField and
republishes.

mikepurvis commented:
I've revised the commit to remove whitespace changes— an overenthusiastic vim plugin. You can make the changes yourself later if you like (for example, with astyle).

Having a separate node is a disadvantage for a few reasons:
- It's not a transparent change, making it less tenable to merge to indigo (users need to launch the adapter node to get legacy support)
- It's a less clean tick-tock, since the user is launching or not-launching the adapter node depending which message they want to use (vs just having a parameter which we can change the default for in Kinetic or perhaps Jade).
- It's less performant, since there's an IPC happening (vs an in-process publish/subscribe).

The adapter being so simple is an argument in my mind toward just baking it in rather than making a separate process (or even a separate class).

mintar commented:
So what about the following behavior:

If `use_magnetic_field_msg == false` (default on =< Indigo):
- subscribe to `geometry_msgs/Vector3Stamped` on topic `imu/mag` (current behavior)
- print warning that default will change in Jade

If `use_magnetic_field_msg == true` (default on >= Jade):
- subscribe to `sensor_msgs/MagneticField` on topic `imu/mag`

... and skip the republishing thing altogether. The user simply sets the param to whatever fits the output of their IMU.

mintar commented:
Alternatively, we could simply add a second topic, let's say `/imu/magnetic_field` of type `sensor_msgs/MagneticField`. No need for a param or any tick-tock, since it's an additional feature, and we don't change the type of the existing topic. Also no need for republishing, we could simply call the other callback directly. This would make it directly obvious to anyone calling `rosnode info` what types of messages the node subscribes to.

P.S.: The reason I don't like the republishing stuff is that it complicates the topic API of the node. As a new user, I'd ask myself "why it is _publishing_ MagneticField messages?".

mikepurvis commented:
Only subscribing to one of them would certainly be the obvious thing; it was just tricky to do cleanly based on how the rest of it is coded, especially in such a way that the legacy code could be easily removed down the road when deemed no longer necessary.

I'm open to an altogether separate topic option, as well, although I feel there should still be just one of them subscribed to at a time (controlled by parameter), to similarly avoid confusion.

mintar commented:
Ok, I get it now, we can't simply call the `imuMagCallback` because of the time synchronizer. I agree now that republishing is the best idea.

> Requires further testing before merge.

Did you get to that yet? If not, I'll do it on Monday (since I'm on vacation right now); otherwise we can merge now.

mintar commented on imu_filter_madgwick/src/imu_filter.cpp:
```diff
@@ -540,4 +560,11 @@ void ImuFilter::reconfigCallback(FilterConfig& config, uint32_t level)
+  mag_msg.header = mag_vector_msg->header;
+  mag_msg.magnetic_field = mag_vector_msg->vector;
+  mag_msg.magnetic_field_covariance[0] = -1;
+  imu_republisher.publish(mag_msg);
```
This should be `mag_republisher_.publish`. Don't bother, I'll fix it.

mintar commented on imu_filter_madgwick/src/imu_filter.cpp:
```diff
@@ -540,4 +560,11 @@ void ImuFilter::reconfigCallback(FilterConfig& config, uint32_t level)
+  MagMsg mag_msg;
+  mag_msg.header = mag_vector_msg->header;
+  mag_msg.magnetic_field = mag_vector_msg->vector;
+  mag_msg.magnetic_field_covariance[0] = -1;
```
This line should be deleted: The MagneticField.msg specifies that...

> A covariance matrix of all zeros will be interpreted as "covariance unknown"
> and to use the data a covariance will have to be assumed or gotten from some
> other source

... which is what we want. You probably mixed this up with Imu.msg, where -1 means "orientation is invalid".

mintar commented:
I couldn't test today, will merge tomorrow (with my tiny fixes) + add a deprecation warning for hydro and indigo + make a fresh release (last one for hydro!).

mintar commented:
Ok, I just tested everything thoroughly and committed a slightly fixed version in 3de491bf. Will release into hydro and indigo now. Many thanks!

mikepurvis commented:
Awesome, thanks!
